chainer/functions/theano/theano_function.py

```diff
@@ -0,0 +1,149 @@
+import collections
+
+import numpy
+import six
+
+try:
+    import theano
+    import theano.sandbox.cuda as theano_cuda
+    _available = True
+except ImportError:
+    _available = False
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _make_var_tuple(vs):
+    msg = ('inputs and outputs must be a TensorVariable, a list '
+           'of TensorVariable or a tuple of TensorVariable')
+
+    if isinstance(vs, theano.tensor.TensorVariable):
+        return vs,
+    elif isinstance(vs, collections.Iterable):
+        vs = tuple(vs)
+        if not all(isinstance(v, theano.tensor.TensorVariable) for v in vs):
+            raise TypeError(msg)
+        return vs
+    else:
+        raise TypeError(msg)
+
+
+class TheanoFunction(function.Function):
+
+    def __init__(self, inputs, outputs, gpu=True):
+        if not _available:
+            msg = '''theano is not installed on your environment.
+Please install theano to activate theano function.
+
+  $ pip install theano'''
+            raise RuntimeError(msg)
+
+        inputs = _make_var_tuple(inputs)
+        outputs = _make_var_tuple(outputs)
+        if gpu:
+            outs = tuple(theano.sandbox.cuda.basic_ops.gpu_from_host(o)
+                         if o.type.dtype == 'float32' else o for o in outputs)
+        else:
+            outs = outputs
+
+        self.func = theano.function(inputs=inputs, outputs=outs)
+        gs = tuple(o.type('g_{}'.format(i)) for i, o in enumerate(outputs))
+        know_grads = dict(zip(outputs, gs))
+
+        grad = theano.tensor.grad(
+            cost=None, wrt=inputs, known_grads=know_grads,
+            disconnected_inputs='ignore')
+
+        if gpu:
+            grad = tuple(theano.sandbox.cuda.basic_ops.gpu_from_host(g)
+                         if g.type.dtype == 'floag32' else g for g in grad)
+
+        self.grad = theano.function(
+            inputs=inputs + gs,
+            outputs=grad,
+            on_unused_input='ignore')
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == len(self.func.input_storage))
+
+        for i, storage in enumerate(self.func.input_storage):
+            expect_type = storage.type
+            actual_type = in_types[i]
```

Comment:
Is `in_types` iterable? If so, `zip(in_types, self.func.input_storage)` is simpler.